File: wiring/wiring.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

//node class
class Node
{
public:
    string name;
    string type;
    int set;
    int place;
};

//edge class
class Edge
{
public:
    Node first;
    Node last;
    int weight;
};

//initialize algorithms
int find(int x, vector<Node> f);
void merge(Node x, Node y);
int compare(const Edge &a, const Edge &b);

//main method
int main()
{
    int n;
    cin >> n;
    int e;
    cin >> e;

    vector<Node> nodes;

    //read in nodes
    for (int i = 0; i < n; i++)
    {
        string label;
        string t;
        cin >> label;
        cin >> t;

        Node n;
        n.name = label;
        n.type = t;
        n.set = 0;
        n.place = 0;

        nodes.push_back(n);
    }

    vector<Edge> edges;

    //read in edges
    for (int i = 0; i < e; i++)
    {
        string f;
        string l;
        int w;
        cin >> f;
        cin >> l;
        cin >> w;

        Edge ed;

        for (int j = 0; j < nodes.size(); j++)
        {
            if (nodes[j].name == f)
            {
                ed.first = nodes[j];
            }
            if (nodes[j].name == l)
            {
                ed.last = nodes[j];
            }
        }

        ed.weight = w;

        edges.push_back(ed);
    }

    //sort edges by weight
    sort(edges.begin(), edges.end(), compare);

    //make array for find union algorithm
    for (int i = 0; i < nodes.size(); i++)
    {
        nodes[i].set = i;
        nodes[i].place = i;
    }

    //Kruskal's Algorithm
    int min = 0;
    int edgesAccepted = 0;
    int index = 0;

    while (edgesAccepted < nodes.size() - 1)
    {
        Edge ed = edges[index];
        int uset = find(ed.first.place, nodes);
        int vset = find(ed.last.place, nodes);
        if (uset != vset)
        {
            edgesAccepted++;
            min += ed.weight;
            merge(ed.first, ed.last);
        }
        index++;
    }

    //print result
    cout << min << endl;

    return 0;
}

//find for find union data structure
int find(int x, vector<Node> f)
{
    for (int i = x; i < f.size(); i++)
    {
        if (f[i].set == i)
        {
            return i;
        }
        i = f[i].set;
    }
}

//union for find union data structure
void merge(Node x, Node y)
{
    y.set = x.set;
}

//for sort by weight
int compare(const Edge &a, const Edge &b)
{
    return (a.weight < b.weight);
}

File: wiring/wiring.java
import java.util.*;

//wiring class
public class wiring{

    //kruskals algorithm
    public static int kruskals(HashMap<String,Node> nodes, Edge[] edges, int edgeNum){
        int edgesAccepted = 0;
        DisjSet s = new DisjSet(nodes.size());
        int nodeNum = nodes.size();
        int i = 0;
        int min = 0;
        
        while(edgesAccepted < nodeNum - 1){
            int uset = s.find(edges[i].first.order);
            int vset = s.find(edges[i].last.order);
            if(uset != vset){
                edgesAccepted++;
                s.union(uset, vset);
                min += edges[i].weight;
                i = edgesAccepted;
            }
            else{
                i++;
            }
        }
        return min;
    }

    // quicksort method
	public static Edge[] quickSort(Edge[] list, int first, int last) {
		if (first < last) {
			Edge temp = list[last];
			int r = (int) (first + (Math.random() * (last - first)));
			list[last] = list[r];
			list[r] = temp;
			int p = partition(list, first, last);
			quickSort(list, first, p - 1);
			quickSort(list, p + 1, last);
		}
		return list;
	}

	// Lomuto's partition for quicksort
	private static int partition(Edge[] list, int first, int last) {
		Edge temp = list[last];
		int prev = first - 1;
		for (int next = first; next <= last - 1; next++) {
			if ((list[next].weight - temp.weight) < 0) {
				prev++;
				Edge pick = list[prev];
				list[prev] = list[next];
				list[next] = pick;
			}
		}
		Edge swap = list[prev + 1];
		list[prev + 1] = list[last];
		list[last] = swap;
		return prev + 1;
	}

    public static void main(String[] args){
        //read in number of nodes and number of edges
        int nodeNum;
        int edgeNum;
        Scanner s = new Scanner(System.in);
        nodeNum = s.nextInt();
        edgeNum = s.nextInt();

        //containers to store nodes and edges
        HashMap<String, Node> map = new HashMap<String, Node>();
        Edge[] edges = new Edge[edgeNum];
        Node prevSwitch = new Node();

        //read in nodes
        for(int i = 0; i < nodeNum; i++){
            Node n = new Node();
            n.name = s.next();
            n.type = s.next();
            n.order = i;
            n.sw = "";
            if(n.type.equals("switch")){
                prevSwitch = n;
                n.sw = n.name;
            }
            else if(n.type.equals("light")){
                n.sw = prevSwitch.name;
            }
            map.put(n.name, n);
        }

        int eNum = 0;
        int sNum = 0;
        Edge[] switches = new Edge[edgeNum];
        //read in edges
        for(int i = 0; i < edgeNum; i++){
            Edge e = new Edge();
            e.first = map.get(s.next());
            e.last = map.get(s.next()); 
            e.weight = Integer.parseInt(s.next());

            //don't add invalid edges to edge list
            if(e.first.type.equals("light") && !(e.last.sw.equals(e.first.sw))){
                ;
            }
            else if(e.last.type.equals("light") && !(e.first.sw.equals(e.last.sw))){
                ;
            } 
            else if(e.first.type.equals("switch") && e.last.type.equals("switch")){
                ;
            }
            else if(e.first.type.equals("switch") || e.last.type.equals("switch") || e.first.type.equals("light") || e.last.type.equals("light")){
                switches[sNum] = e;
                sNum++;
            }
            else{
                edges[eNum] = e;
                eNum++;
            }
        }

        //sort edges by weight
        quickSort(edges, 0, eNum - 1);
        quickSort(switches, 0, sNum - 1);

        for(int i = 0; i < sNum; i++){
            edges[eNum] = switches[i];
            eNum++;
        }

        System.out.println(kruskals(map, edges, eNum));
    }
}

